pkg/splunk/enterprise/util.go

```diff
@@ -292,3 +294,236 @@ func DeleteOwnerReferencesForS3SecretObjects(client splcommon.ControllerClient,
+		appDeployContext.AppsSrcDeployStatus[appSrc] = appSrcDeploymentInfo
+	}
+
+	// 2. Go through each AppSrc from the remote listing
+	for appSrc, s3Response := range remoteObjListingMap {
+		// 2.1 Mark Apps for deletion if they are missing in remote listing
+		appSrcDeploymentInfo, appSrcExistsLocally := appDeployContext.AppsSrcDeployStatus[appSrc]
+
+		if appSrcExistsLocally {
+			currentList := appSrcDeploymentInfo.AppDeploymentInfoList
+			for appIdx := range currentList {
+				if !checkIfAnAppIsActiveOnRemoteStore(currentList[appIdx].AppName, s3Response.Objects) {
+					setStateAndStatusForAppDeployInfo(currentList[appIdx], enterprisev1.RepoStateDeleted, enterprisev1.DeployStatusPending)
+					appDeployContext.IsDeploymentInProgress = true
+				}
+			}
+		}
+
+		// 2.2 Check for any App changes(Ex. A new App source, a new App added/updated)
+		if createOrUpdateAppSrcDeploymentInfo(&appSrcDeploymentInfo, s3Response.Objects) {
+			appDeployContext.IsDeploymentInProgress = true
+		}
+
+		// Finally update the Map entry with latest info
+		appDeployContext.AppsSrcDeployStatus[appSrc] = appSrcDeploymentInfo
+	}
+
+	return true, err
+}
+
+// isAppExtentionValid checks if an app extention is supported or not
+func isAppExtentionValid(receivedKey string) bool {
+	appExtIdx := strings.LastIndex(receivedKey, ".")
+	if appExtIdx < 0 {
+		return false
+	}
+
+	switch appExt := receivedKey[appExtIdx+1:]; appExt {
+	case "spl":
+		return true
+
+	case "tgz":
+		return true
+
+	default:
+		return false
+	}
+}
+
+// createOrUpdateAppSrcDeploymentInfo  modifies the App deployment status as perceived from the remote object listing
+func createOrUpdateAppSrcDeploymentInfo(appSrcDeploymentInfo *enterprisev1.AppSrcDeployInfo, remoteS3ObjList []*splclient.RemoteObject) bool {
+	scopedLog := log.WithName("createOrUpdateAppSrcDeploymentInfo").WithValues("Called with length: ", len(remoteS3ObjList))
+
+	var found bool
+	var appName string
+	var newAppInfoList []enterprisev1.AppDeploymentInfo
+	var appChangesDetected bool
+	var appDeployInfo enterprisev1.AppDeploymentInfo
+
+	for _, remoteObj := range remoteS3ObjList {
+		receivedKey := *remoteObj.Key
+		if !isAppExtentionValid(receivedKey) {
+			scopedLog.Info("App name Parsig", " Ignoring the key: %s, with invalid extention for the app", receivedKey)
+			continue
+		}
+
+		nameAt := strings.LastIndex(receivedKey, "/")
+		if nameAt == -1 {
+			appName = receivedKey
+		} else {
+			appName = receivedKey[nameAt+1:]
+		}
+
+		// Now update App status as seen in the remote listing
+		found = false
+		for _, appDeployInfo = range appSrcDeploymentInfo.AppDeploymentInfoList {
+			if appDeployInfo.AppName == appName {
+				found = true
+				if appDeployInfo.ObjectHash != *remoteObj.Etag {
+					appDeployInfo.ObjectHash = *remoteObj.Etag
+					appDeployInfo.DeployStatus = enterprisev1.DeployStatusPending
+					appChangesDetected = true
+				} else if appDeployInfo.RepoState == enterprisev1.RepoStateDeleted {
```

Comment:
How about this scenario:
1. App is removed from remoteStore -  so the `RepoState` is set to `RepoStateDeleted`. I guess rest of the App metadata (such as `ObjectHash`) is preserved
2. App is updated offline & then copied back to remoteStore - as per this logic `DeployStatus` is set to `DeployStatusPending` but `RepoState` will never be set to `RepoStateActive`. Also `Size` isn't updated either

One way of handling I can think of:
```
			if appDeployInfo.AppName == appName {
				found = true
				// App has either been updated or enabled back
				if appDeployInfo.ObjectHash != *remoteObj.Etag ||  appDeployInfo.RepoState == enterprisev1.RepoStateDeleted {
					scopedLog.Info("App %s has either been updated or enabled back", appName)
					appDeployInfo.ObjectHash = *remoteObj.Etag
					appDeployInfo.Size = uint64(*remoteObj.Size)
					appDeployInfo.DeployStatus = enterprisev1.DeployStatusPending
					appDeployInfo.RepoState = enterprisev1.RepoStateActive
					appChangesDetected = true
				} 
				break
```

Reply:
Line 481 sets the repoState to Active again. Size, Ownership, last modified fields we are intentionally not adding, for better scaling. But these are printed as part of the logs and can be checked on remote stores at any time.

Reply:
Sorry, I overlooked the scenario. Will address this case more wholistically.